File: scripts/compile-contract.py
#!/usr/bin/env python3
"""
Скрипт для компиляции Solidity контрактов и генерации TypeScript конфигурации
"""

import json
import os
import subprocess
import sys
from pathlib import Path

# Конфигурация
CONTRACT_FILE = "contracts/PongTournamentScores.sol"
OUTPUT_FILE = "src/blockchain/contractConfig.ts"
BUILD_DIR = "build"

def ensure_solc_installed():
    """Проверяет и устанавливает solc если необходимо"""
    try:
        subprocess.run(["solc", "--version"], capture_output=True, check=True)
        print("✓ Solidity compiler найден")
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        print("⚠ Solidity compiler не найден")
        return install_solc()

def install_solc():
    """Устанавливает solc через pip"""
    try:
        print("📦 Устанавливаю py-solc-x...")
        subprocess.run([sys.executable, "-m", "pip", "install", "py-solc-x"], check=True)
        
        # Импортируем и устанавливаем solc
        from solcx import install_solc, set_solc_version
        
        print("📦 Устанавливаю Solidity 0.8.19...")
        install_solc('0.8.19')
        set_solc_version('0.8.19')
        
        print("✓ Solidity compiler установлен")
        return True
    except Exception as e:
        print(f"❌ Ошибка установки solc: {e}")
        return False

def compile_contract():
    """Компилирует контракт и возвращает ABI и байткод"""
    print("🔨 Компилирую контракт...")
    
    try:
        from solcx import compile_files
        
        # Компилируем файл
        compiled = compile_files([CONTRACT_FILE])
        
        # Получаем ключ контракта (например: contracts/PongTournamentScores.sol:PongTournamentScores)
        contract_key = None
        for key in compiled.keys():
            if "PongTournamentScores" in key:
                contract_key = key
                break
        
        if not contract_key:
            raise Exception("Контракт PongTournamentScores не найден в скомпилированных файлах")
        
        contract = compiled[contract_key]
        abi = contract['abi']
        bytecode = '0x' + contract['bin']
        
        print("✓ Контракт скомпилирован")
        return abi, bytecode
        
    except ImportError:
        print("❌ py-solc-x не установлен. Пытаюсь использовать системный solc...")
        return compile_with_system_solc()
    except Exception as e:
        print(f"❌ Ошибка компиляции: {e}")
        return None, None

def compile_with_system_solc():
    """Компилирует контракт используя системный solc"""
    try:
        # Создаем папку build
        os.makedirs(BUILD_DIR, exist_ok=True)
        
        # Компилируем контракт
        cmd = [
            "solc",
            "--optimize",
            "--abi",
            "--bin",
            "--output-dir", BUILD_DIR,
            CONTRACT_FILE
        ]
        
        subprocess.run(cmd, check=True, capture_output=True)
        
        # Читаем результаты
        abi_file = os.path.join(BUILD_DIR, "PongTournamentScores.abi")
        bin_file = os.path.join(BUILD_DIR, "PongTournamentScores.bin")
        
        if not os.path.exists(abi_file) or not os.path.exists(bin_file):
            raise Exception("Скомпилированные файлы не найдены")
        
        with open(abi_file, 'r') as f:
            abi = json.load(f)
        
        with open(bin_file, 'r') as f:
            bytecode = '0x' + f.read().strip()
        
        return abi, bytecode
        
    except Exception as e:
        print(f"❌ Ошибка компиляции с системным solc: {e}")
        return None, None

def generate_typescript_config(abi, bytecode):
    """Генерирует TypeScript конфигурационный файл"""
    content = f"""// Автоматически сгенерировано из {CONTRACT_FILE}
// Не редактируйте этот файл вручную - используйте npm run compile-contract

export const PongTournamentScoresABI = {json.dumps(abi, indent=4)};

export const PongTournamentScoresBytecode = "{bytecode}";
"""
    
    try:
        # Создаем директорию если её нет
        os.makedirs(os.path.dirname(OUTPUT_FILE), exist_ok=True)
        
        with open(OUTPUT_FILE, 'w') as f:
            f.write(content)
        
        print(f"✓ Конфигурация сохранена в {OUTPUT_FILE}")
        return True
        
    except Exception as e:
        print(f"❌ Ошибка записи файла: {e}")
        return False

def main():
    """Основная функция"""
    print("🚀 Генерация ABI и байткода для смарт-контракта")
    print("=" * 50)
    
    # Проверяем наличие контракта
    if not os.path.exists(CONTRACT_FILE):
        print(f"❌ Файл контракта не найден: {CONTRACT_FILE}")
        sys.exit(1)
    
    # Проверяем и устанавливаем solc
    if not ensure_solc_installed():
        sys.exit(1)
    
    # Компилируем контракт
    abi, bytecode = compile_contract()
    if not abi or not bytecode:
        sys.exit(1)
    
    # Генерируем TypeScript конфигурацию
    if not generate_typescript_config(abi, bytecode):
        sys.exit(1)
    
    print("=" * 50)
    print("✅ Генерация завершена успешно!")
    print(f"📁 ABI и байткод обновлены в {OUTPUT_FILE}")

if __name__ == "__main__":
    main()

File: scripts/compile-contract.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Конфигурация
const CONTRACT_FILE = 'contracts/PongTournamentScores.sol';
const OUTPUT_FILE = 'src/blockchain/contractConfig.ts';
const BUILD_DIR = 'build';

// Проверка и установка solc
function ensureSolcInstalled() {
    try {
        execSync('solc --version', { stdio: 'pipe' });
        console.log('✓ Solidity compiler найден');
        return true;
    } catch (error) {
        console.log('⚠ Solidity compiler не найден. Устанавливаю локально...');
        try {
            // Устанавливаем solc локально в проект
            execSync('npm install solc@0.8.19', { stdio: 'inherit' });
            console.log('✓ Solidity compiler установлен локально');
            return true;
        } catch (installError) {
            console.error('❌ Ошибка установки solc:', installError.message);
            return false;
        }
    }
}

// Компиляция контракта
function compileContract() {
    console.log('🔨 Компилирую контракт...');

    try {
        // Пробуем использовать локальный solc из node_modules
        const localSolc = path.join('node_modules', '.bin', 'solc');
        let solcCommand = 'solc';

        if (fs.existsSync(localSolc)) {
            solcCommand = localSolc;
            console.log('Используется локальный solc');
        }

        // Создаем папку build
        if (!fs.existsSync(BUILD_DIR)) {
            fs.mkdirSync(BUILD_DIR, { recursive: true });
        }

        // Компилируем контракт
        const command = `${solcCommand} --optimize --abi --bin --output-dir ${BUILD_DIR} ${CONTRACT_FILE}`;
        execSync(command, { stdio: 'pipe' });

        console.log('✓ Контракт скомпилирован');
        return true;

    } catch (error) {
        // Если системный/локальный solc не работает, пробуем через Node.js API
        console.log('⚠ Системный solc не работает, пробую Node.js API...');
        return compileWithNodeAPI();
    }
}

// Компиляция через Node.js API
function compileWithNodeAPI() {
    try {
        const solc = require('solc');

        // Читаем исходный код контракта
        const source = fs.readFileSync(CONTRACT_FILE, 'utf8');

        // Настройки компиляции
        const input = {
            language: 'Solidity',
            sources: {
                'PongTournamentScores.sol': {
                    content: source
                }
            },
            settings: {
                outputSelection: {
                    '*': {
                        '*': ['abi', 'evm.bytecode']
                    }
                },
                optimizer: {
                    enabled: true,
                    runs: 200
                }
            }
        };

        // Компилируем
        const compiled = JSON.parse(solc.compile(JSON.stringify(input)));

        // Проверяем ошибки
        if (compiled.errors) {
            const errors = compiled.errors.filter(error => error.severity === 'error');
            if (errors.length > 0) {
                console.error('❌ Ошибки компиляции:');
                errors.forEach(error => console.error(error.formattedMessage));
                return false;
            }
        }

        // Извлекаем результаты
        const contract = compiled.contracts['PongTournamentScores.sol']['PongTournamentScores'];
        const abi = contract.abi;
        const bytecode = '0x' + contract.evm.bytecode.object;

        // Создаем папку build и сохраняем файлы
        if (!fs.existsSync(BUILD_DIR)) {
            fs.mkdirSync(BUILD_DIR, { recursive: true });
        }

        fs.writeFileSync(path.join(BUILD_DIR, 'PongTournamentScores.abi'), JSON.stringify(abi, null, 2));
        fs.writeFileSync(path.join(BUILD_DIR, 'PongTournamentScores.bin'), contract.evm.bytecode.object);

        console.log('✓ Контракт скомпилирован через Node.js API');
        return true;

    } catch (error) {
        console.error('❌ Ошибка компиляции через Node.js API:', error.message);
        return false;
    }
}

// Чтение скомпилированных файлов
function readCompiledFiles() {
    const abiFile = path.join(BUILD_DIR, 'PongTournamentScores.abi');
    const binFile = path.join(BUILD_DIR, 'PongTournamentScores.bin');

    if (!fs.existsSync(abiFile) || !fs.existsSync(binFile)) {
        console.error('❌ Скомпилированные файлы не найдены');
        return null;
    }

    try {
        const abi = JSON.parse(fs.readFileSync(abiFile, 'utf8'));
        const bytecode = '0x' + fs.readFileSync(binFile, 'utf8').trim();

        return { abi, bytecode };
    } catch (error) {
        console.error('❌ Ошибка чтения файлов:', error.message);
        return null;
    }
}

// Генерация TypeScript конфигурации
function generateTypeScriptConfig(abi, bytecode) {
    const timestamp = new Date().toISOString();
    const content = `// Автоматически сгенерировано из ${CONTRACT_FILE}
// Дата генерации: ${timestamp}
// Не редактируйте этот файл вручную - используйте npm run compile-contract

export const PongTournamentScoresABI = ${JSON.stringify(abi, null, 4)};

export const PongTournamentScoresBytecode = "${bytecode}";
`;

    try {
        // Создаем директорию если её нет
        const dir = path.dirname(OUTPUT_FILE);
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
        }

        fs.writeFileSync(OUTPUT_FILE, content, 'utf8');
        console.log(`✓ Конфигурация сохранена в ${OUTPUT_FILE}`);
        return true;
    } catch (error) {
        console.error('❌ Ошибка записи файла:', error.message);
        return false;
    }
}

// Основная функция
function main() {
    console.log('🚀 Компиляция смарт-контракта Solidity');
    console.log('='.repeat(50));

    // Проверяем наличие контракта
    if (!fs.existsSync(CONTRACT_FILE)) {
        console.error(`❌ Файл контракта не найден: ${CONTRACT_FILE}`);
        process.exit(1);
    }

    // Проверяем и устанавливаем solc
    if (!ensureSolcInstalled()) {
        process.exit(1);
    }

    // Компилируем контракт
    if (!compileContract()) {
        process.exit(1);
    }

    // Читаем результаты компиляции
    const compiled = readCompiledFiles();
    if (!compiled) {
        process.exit(1);
    }

    // Генерируем TypeScript конфигурацию
    if (!generateTypeScriptConfig(compiled.abi, compiled.bytecode)) {
        process.exit(1);
    }

    console.log('='.repeat(50));
    console.log('✅ Компиляция завершена успешно!');
    console.log(`📁 ABI и байткод обновлены в ${OUTPUT_FILE}`);
    console.log(`🗂️  Промежуточные файлы сохранены в ${BUILD_DIR}/`);
}

// Запускаем скрипт
if (require.main === module) {
    main();
}

module.exports = { main };
